### ISSUE TITLE
Remove "warning" when package metadata generate from `meson.build`

### ISSUE THREAD

**Issue by dnicolodi** [mesonbuild/meson-python#306]
Currently a "warning" (some yellow text prefixed with an exclamation mark, which is very hard to read with my terminal emulator settings, but that is probably a problem of the default palette of the terminal emular) is emitted when the package metadata is synthesized from the project metadata in `meson.build`.

I think we should either decide that this is supported and make the warning simply an informative message, or decide that it is a bad idea and error out. As we can get the minimum required information to build a package out of `meson.build` we should support this and remove the warning. It seems its main purpose is to inform the user that they may have a typo in their `pyproject.toml` but the warning is very easy to miss within the rest of the output. The effects of not populating the metadata from `pyproject.toml` are probably much more visible.

Also, having all out test packages emitting what looks like a warning looks like poor form to me.

**Comment by rgommers** [comment]
What can actually be taken from `meson.build`? Just name and version, or more? `license` is another keyword in Meson's `project` that matches, but that's going to be ambiguous once we support PEP 639.

**Comment by dnicolodi** [comment]
Just name and version.

The PEP 621 `license` field is for a license text, while the `licence` parameter to the Meson's `project()` is a license identifier. It is recommended to be a SPDX license expression, but it is not enforced. I'm not very familiar with PEP 639, but my understanding is that it will allow to use SPDX license expressions in Python packages metadata. Therefore, once PEP 639 is settled, approved, and implemented we could map the Meson's project license to Python package metadata.

**Comment by rgommers** [comment]
Given that a `pyproject.toml` file is required anyway and the gain of supporting a couple of fields from `meson.build` is very small, I'd personally go for requiring PEP 621 metadata. 

However, if others want to support using `name` and `version` from `meson.build`, that's fine with me too - in that case I do agree that the warning should be removed, and the extent of the support documented in our html docs.

**Comment by eli-schwartz** [comment]
You *have* to use the version scraped from meson.build if you want to support git-based versions, for example by running `python -m setuptools_scm` to print the version and read that into meson.build. It's also potentially useful for buildsystem processing, so you may want to single-source it rather than defining it in multiple places including meson.build as well as pyproject.toml.

It's probably the single most common metadata field to be specified as dynamic.

Name cannot be dynamic one way or the other except by totally opting out of PEP 621 altogether, which seems unlikely. And it's questionable what other metadata fields are super interesting to define dynamically...

... although I note that Meson supports both `license: ...` name (assumed to be SPDX identifiers) and `license_files: ` -- the latter can be installed with `meson setup --licensedir /path/to/licenses/directory` and also generate an SBOM manifest referencing both the license name and paths to license files.

> The PEP 621 `license` field is for a license text, while the `licence` parameter to the Meson's `project()` is a license identifier. It is recommended to be a SPDX license expression, but it is not enforced. I'm not very familiar with PEP 639, but my understanding is that it will allow to use SPDX license expressions in Python packages metadata. Therefore, once PEP 639 is settled, approved, and implemented we could map the Meson's project license to Python package metadata.

The PEP 621 license field is explicitly meaningless. It permits two types of values:
- file, indicating the path to a license file
- text, which must be identical to the Core Metadata meaning. The Core Metadata describes license as an alternative for Trove classifiers for cases where there isn't a Trove classifier

These are fundamentally incompatible definitions, you cannot pack them both into the same Core Metadata. I would argue that it's an illegal violation of the Core Metadata to actually use PEP 621 licenses with the `file` type of value.

You (rfc 2119) MUST NOT include the text of a license file as Core Metadata "License", regardless of any other considerations.

Yes, it's unfortunate that PEP 621 left this footgun lying around, but there you have it.

It's even more unfortunate that despite the fact that the ***logical*** way to use this metadata is to have both an SPDX identifier text value and a file value, PEP 621 goes out of its way to tell you that you're utterly forbidden from using both at the same time.

> These keys are mutually exclusive, so a tool MUST raise an error if the metadata specifies both keys. 

IMHO the best option here, simply to resolve a terrible case of the ambiguities, is to stick to Meson's license handling which has clear semantics: one field for SPDX identifiers, one field describing the path to a file or files, and it's a good idea to use both fields. Use the former to dynamically fill out the Core Metadata, use the latter to install files to a subdirectory of, probably, the dist-info.

**Comment by dnicolodi** [comment]
> You _have_ to use the version scraped from meson.build if you want to support git-based versions, for example by running `python -m setuptools_scm` to print the version and read that into meson.build. It's also potentially useful for buildsystem processing, so you may want to single-source it rather than defining it in multiple places including meson.build as well as pyproject.toml.

Yes. This is supported declaring the `version` field in `pyproject.toml` to be dyamic:
```toml
[project]
name = foo
dynamic = ['version']
```
When meson-python encounters that, it reads the version from the Meson metadata.

> IMHO the best option here, simply to resolve a terrible case of the ambiguities, is to stick to Meson's license handling which has clear semantics: one field for SPDX identifiers, one field describing the path to a file or files, and it's a good idea to use both fields. Use the former to dynamically fill out the Core Metadata, use the latter to install files to a subdirectory of, probably, the dist-info.

I think we should continue to do what we do now (handle the `license` filed in `pyproject.toml` as prescribed in PEP 621) and fix the mess once PEP 639 is approved and implemented in the rest of the packaging stack.

> although I note that Meson supports both `license: ...` name (assumed to be SPDX identifiers) and `license_files: `

Someone must have forgotten to document this 😄 . I remember seeing the PR that implemented it, but I could not tell if it was merged because `project()` documentation does not mention it https://mesonbuild.com/Reference-manual_functions.html#project Although I see a reference to the concept here https://mesonbuild.com/Reference-manual_builtin_meson.html#mesonproject_license_files

**Comment by eli-schwartz** [comment]
> Someone must have forgotten to document this smile .

I resemble that remark...

Yes, I updated half the documentation but forgot the other half. My apologies.

**Comment by rgommers** [comment]
> When meson-python encounters that, it reads the version from the Meson introspection data. If there is no `project` section in `pyproject.toml` we read `name` and `version` from the Meson introspection data.

Yes, so we support it. What I meant was that we probably don't need two different ways to support it, just to save a few characters in `pyproject.toml`. Plus explicit is better than implicit.

**Comment by dnicolodi** [comment]
I like the fact that a minimal package does not require duplicating the same information in two places (`pyproject.toml` `[project]` section and `meson.build` `project()`) therefore I would like to keep the support we already have and just paint the notice in a different color.

**Comment by FFY00** [comment]
I implemented it because it's useful when you want to quickly package a Meson project as a Python wheel, but I acknowledge that is a very small use-case. If you want to remove it, that's fine.

**Comment by dnicolodi** [comment]
I would like to keep the support. Just tone down a bit the message printed on the console.

**Comment by dnicolodi** [comment]
It seems that support for a missing `[project]` section in `pyproject.toml` is mandated (or at least allowed, the wording is not clear) by PEP 621 https://peps.python.org/pep-0621/#table-name:
> The lack of a [project] table implicitly means the build back-end will dynamically provide all fields.
